planner/cascades/implementation.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	plannercore "github.com/pingcap/tidb/planner/core"
+)
+
+// Implementation defines the interface for cost of physical plan.
+type Implementation interface {
+	calcCost(outCount float64, childCosts []float64, children ...*Group) float64
```

Comment:
I wonder that `Implementation` seems like the current struct `task`. So how will the `calcCost` method work? I think the `cost` depends on the specific physical operator?
We'll let `PhysicalPlan` implement one more method with the same function signature?

Reply:
Here is an example of `calcCost` for hash join and selection, the cost computations are split to next PR.
```golang
func (impl *HashJoinImpl) calcCost(outCount float64, childCosts []float64, children ...*Group) float64 {
    lCnt := math.Min(children[0].prop.Stats.RowCount, impl.plan.GetChildReqProps(0).ExpectedCnt)
    rCnt := math.Min(children[1].prop.Stats.RowCount, impl.plan.GetChildReqProps(1).ExpectedCnt)
    hashJoin := impl.plan.(*plannercore.PhysicalHashJoin)
    impl.cost = hashJoin.GetCost(lCnt, rCnt) + childCosts[0] + childCosts[1]
    return impl.cost
}

func (impl *SelectionImpl) calcCost(outCount float64, childCosts []float64, children ...*Group) float64 {
    cnt := math.Min(children[0].prop.Stats.RowCount, impl.plan.GetChildReqProps(0).ExpectedCnt)
    impl.cost = cnt*plannercore.CPUFactor + childCosts[0]
    return impl.cost
}
```
We implement `calcCost` in `Implementation` so `PhysicalPlan` would not be modified.

Reply:
got